plugins/movetoarchive.koplugin/main.lua

```diff
@@ -0,0 +1,161 @@
+            {
+                text = _("Set archive directory"),
+                keep_menu_open = true,
+                callback =  self.setArchiveDirectory,
+            }
+        },
+    }
+end
+
+function MoveToArchive:moveToArchive()
+    local move_done_text = _("Book moved.\nDo you want to open it from the archive folder?")
+    self:commonProcess(true, move_done_text)
+end
+
+function MoveToArchive:copyToArchive()
+    local copy_done_text =_("Book copied.\nDo you want to open it from the archive folder?")
+    self:commonProcess(false, copy_done_text)
+end
+
+function MoveToArchive:commonProcess(is_move_process, moved_done_text)
+    if not self.archive_dir_path then
+        self:showNoArchiveConfirmBox()
+        return
+    end
+    local document_full_path = self.ui.document.file
+    local filename
+    self.last_copied_from_dir, filename = util.splitFilePathName(document_full_path)
+
+    logger.dbg("MoveToArchive: last_moved/copied_from_dir :", self.last_copied_from_dir)
+    self.move_to_archive_settings:saveSetting("last_copied_from_dir", ("%s/"):format(self.last_copied_from_dir))
+
+    self.ui:onClose()
+    if is_move_process then
+        FileManager:moveFile(document_full_path, self.archive_dir_path)
+        FileManager:moveFile(DocSettings:getSidecarDir(document_full_path), self.archive_dir_path)
+    else
+        FileManager:copyFileFromTo(document_full_path, self.archive_dir_path)
+        FileManager:copyRecursive(DocSettings:getSidecarDir(document_full_path), self.archive_dir_path)
+    end
+    local dest_file = string.format("%s/%s", self.archive_dir_path, filename)
+    ReadHistory:updateItemByPath(document_full_path, dest_file)
+    ReadCollection:updateItemByPath(document_full_path, dest_file)
+    -- Update last open file.
+    if G_reader_settings:readSetting("lastfile") == document_full_path then
+        G_reader_settings:saveSetting("lastfile", dest_file)
+    end
+
+    UIManager:show(ConfirmBox:new{
+    text = moved_done_text,
+        ok_callback = function ()
+            ReaderUI:showReader(self.archive_dir_path .. filename)
+        end,
+        cancel_callback = function ()
+            if DocSettings:open(dest_file).data.percent_finished == 1 then
+                ReadHistory:removeItemByPath(dest_file)
+            end
+            self:openFileBrowser(self.last_copied_from_dir)
+        end,
+    })
+end
+
+
+function MoveToArchive:setArchiveDirectory()
+    require("ui/downloadmgr"):new{
+        onConfirm = function(path)
+            self.archive_dir_path = path
+            self.move_to_archive_settings:saveSetting("archive_dir", ("%s/"):format(path))
+            self.move_to_archive_settings:flush()
+        end,
+    }:chooseDir()
+end
+
+function MoveToArchive:showNoArchiveConfirmBox()
+    UIManager:show(ConfirmBox:new{
+        text = _("No archive directory.\nDo you want to set it now?"),
+        ok_text = _("Set archive folder"),
+        ok_callback = self.setArchiveDirectory,
```

Comment:
Again not using `:`, so not providing `self` to `MoveToArchive:setArchiveDirectory()`, where you expect it.
`ok_callback = function() self:setArchiveDirectory() end,`